File: python/unified_channel/adapters/matrix.py
"""Matrix adapter — bridges matrix-nio to UnifiedMessage.

Requires: pip install matrix-nio[e2e]
"""

from __future__ import annotations

import asyncio
import logging
from datetime import datetime
from typing import AsyncIterator

from nio import (
    AsyncClient,
    LoginResponse,
    RoomMessageText,
    RoomMessageImage,
    RoomMessageVideo,
    InviteMemberEvent,
)

from ..adapter import ChannelAdapter
from ..types import (
    ChannelStatus,
    ContentType,
    Identity,
    MessageContent,
    OutboundMessage,
    UnifiedMessage,
)

logger = logging.getLogger(__name__)


class MatrixAdapter(ChannelAdapter):
    """
    Matrix channel adapter using matrix-nio.

    Supports E2E encryption if matrix-nio[e2e] is installed.
    """

    channel_id = "matrix"

    def __init__(
        self,
        homeserver: str,
        user_id: str,
        password: str | None = None,
        access_token: str | None = None,
        *,
        allowed_room_ids: set[str] | None = None,
        auto_join: bool = True,
        command_prefix: str = "/",
        device_name: str = "unified-channel",
    ) -> None:
        self._homeserver = homeserver
        self._user_id = user_id
        self._password = password
        self._access_token = access_token
        self._allowed_rooms = allowed_room_ids
        self._auto_join = auto_join
        self._prefix = command_prefix
        self._device_name = device_name
        self._queue: asyncio.Queue[UnifiedMessage] = asyncio.Queue()
        self._connected = False
        self._last_activity: datetime | None = None
        self._client: AsyncClient | None = None

    async def connect(self) -> None:
        self._client = AsyncClient(self._homeserver, self._user_id)

        if self._access_token:
            self._client.access_token = self._access_token
        elif self._password:
            resp = await self._client.login(self._password, device_name=self._device_name)
            if not isinstance(resp, LoginResponse):
                raise RuntimeError(f"matrix login failed: {resp}")
            logger.info("matrix logged in: %s", self._user_id)
        else:
            raise ValueError("matrix: either password or access_token required")

        # Register callbacks
        self._client.add_event_callback(self._on_message, RoomMessageText)
        self._client.add_event_callback(self._on_media, (RoomMessageImage, RoomMessageVideo))
        if self._auto_join:
            self._client.add_event_callback(self._on_invite, InviteMemberEvent)

        # Start sync in background
        self._connected = True
        asyncio.create_task(self._sync_loop())
        logger.info("matrix connected: %s @ %s", self._user_id, self._homeserver)

    async def _sync_loop(self) -> None:
        try:
            await self._client.sync_forever(timeout=30000, full_state=True)
        except Exception as e:
            logger.error("matrix sync error: %s", e)
            self._connected = False

    async def disconnect(self) -> None:
        if self._client:
            await self._client.close()
        self._connected = False
        logger.info("matrix disconnected")

    async def receive(self) -> AsyncIterator[UnifiedMessage]:
        while self._connected:
            try:
                msg = await asyncio.wait_for(self._queue.get(), timeout=1.0)
                yield msg
            except asyncio.TimeoutError:
                continue

    async def send(self, msg: OutboundMessage) -> str | None:
        if not self._client:
            raise RuntimeError("matrix not connected")

        content = {"msgtype": "m.text", "body": msg.text}
        if msg.parse_mode == "html":
            content["format"] = "org.matrix.custom.html"
            content["formatted_body"] = msg.text

        resp = await self._client.room_send(
            room_id=msg.chat_id,
            message_type="m.room.message",
            content=content,
        )
        self._last_activity = datetime.now()
        return getattr(resp, "event_id", None)

    async def get_status(self) -> ChannelStatus:
        return ChannelStatus(
            connected=self._connected,
            channel="matrix",
            account_id=self._user_id,
            last_activity=self._last_activity,
        )

    async def _on_invite(self, room, event) -> None:
        if event.membership == "invite" and event.state_key == self._user_id:
            await self._client.join(room.room_id)
            logger.info("matrix: auto-joined %s", room.room_id)

    async def _on_message(self, room, event: RoomMessageText) -> None:
        # Skip own messages
        if event.sender == self._user_id:
            return
        if self._allowed_rooms and room.room_id not in self._allowed_rooms:
            return

        text = event.body
        self._last_activity = datetime.now()

        if text.startswith(self._prefix):
            parts = text[len(self._prefix):].split()
            cmd = parts[0] if parts else ""
            args = parts[1:]
            mc = MessageContent(type=ContentType.COMMAND, text=text, command=cmd, args=args)
        else:
            mc = MessageContent(type=ContentType.TEXT, text=text)

        msg = UnifiedMessage(
            id=event.event_id,
            channel="matrix",
            sender=Identity(
                id=event.sender,
                display_name=room.user_name(event.sender),
            ),
            content=mc,
            timestamp=datetime.fromtimestamp(event.server_timestamp / 1000),
            chat_id=room.room_id,
            raw=event,
        )
        await self._queue.put(msg)

    async def _on_media(self, room, event) -> None:
        if event.sender == self._user_id:
            return
        if self._allowed_rooms and room.room_id not in self._allowed_rooms:
            return

        self._last_activity = datetime.now()
        mtype = "image" if isinstance(event, RoomMessageImage) else "video"
        url = getattr(event, "url", None)

        mc = MessageContent(
            type=ContentType.MEDIA,
            text=getattr(event, "body", ""),
            media_url=url,
            media_type=mtype,
        )
        msg = UnifiedMessage(
            id=event.event_id,
            channel="matrix",
            sender=Identity(
                id=event.sender,
                display_name=room.user_name(event.sender),
            ),
            content=mc,
            timestamp=datetime.fromtimestamp(event.server_timestamp / 1000),
            chat_id=room.room_id,
            raw=event,
        )
        await self._queue.put(msg)

File: rust/src/adapters/matrix.rs
//! Matrix adapter using the matrix-sdk crate.

use async_trait::async_trait;
use chrono::Utc;
use std::sync::Arc;
use tokio::sync::Mutex;

use matrix_sdk::{
    config::SyncSettings,
    room::Room,
    ruma::{
        events::room::message::{
            MessageType, OriginalSyncRoomMessageEvent, RoomMessageEventContent,
        },
        OwnedRoomId, RoomId,
    },
    Client,
};

use crate::adapter::{ChannelAdapter, MessageHandler};
use crate::error::{Error, Result};
use crate::types::*;

/// Matrix adapter configuration.
pub struct MatrixConfig {
    /// Homeserver URL (e.g., "https://matrix.org").
    pub homeserver_url: String,
    /// Username (localpart, e.g., "bot" for @bot:matrix.org).
    pub username: String,
    /// Password.
    pub password: String,
    /// Optional device display name.
    pub device_name: Option<String>,
}

/// Matrix adapter backed by matrix-sdk.
pub struct MatrixAdapter {
    config: MatrixConfig,
    connected: bool,
    last_activity: Option<chrono::DateTime<Utc>>,
    user_id: Option<String>,
    handler: Arc<Mutex<Option<MessageHandler>>>,
    client: Option<Client>,
}

impl MatrixAdapter {
    /// Create a new Matrix adapter.
    pub fn new(config: MatrixConfig) -> Self {
        Self {
            config,
            connected: false,
            last_activity: None,
            user_id: None,
            handler: Arc::new(Mutex::new(None)),
            client: None,
        }
    }
}

#[async_trait]
impl ChannelAdapter for MatrixAdapter {
    fn channel_id(&self) -> &str {
        "matrix"
    }

    async fn connect(&mut self) -> Result<()> {
        let client = Client::builder()
            .homeserver_url(&self.config.homeserver_url)
            .build()
            .await
            .map_err(|e| Error::Connection(format!("matrix client build: {}", e)))?;

        client
            .matrix_auth()
            .login_username(&self.config.username, &self.config.password)
            .device_display_name(
                self.config
                    .device_name
                    .as_deref()
                    .unwrap_or("unified-channel"),
            )
            .await
            .map_err(|e| Error::Connection(format!("matrix login: {}", e)))?;

        self.user_id = client.user_id().map(|id| id.to_string());
        self.client = Some(client.clone());

        // Register message handler
        let handler = Arc::clone(&self.handler);
        client.add_event_handler(
            move |event: OriginalSyncRoomMessageEvent, room: Room| {
                let handler = Arc::clone(&handler);
                async move {
                    let h = handler.lock().await;
                    if let Some(ref callback) = *h {
                        let text = match &event.content.msgtype {
                            MessageType::Text(t) => t.body.clone(),
                            MessageType::Notice(n) => n.body.clone(),
                            _ => return,
                        };

                        let unified = UnifiedMessage {
                            id: event.event_id.to_string(),
                            channel: "matrix".to_string(),
                            sender: Identity::new(event.sender.to_string()),
                            content: MessageContent::text(text),
                            timestamp: Utc::now(),
                            thread_id: None,
                            reply_to_id: None,
                            chat_id: Some(room.room_id().to_string()),
                            raw: None,
                            metadata: Default::default(),
                        };
                        callback(unified);
                    }
                }
            },
        );

        // Start sync in background
        let sync_client = client.clone();
        tokio::spawn(async move {
            sync_client.sync(SyncSettings::default()).await;
        });

        self.connected = true;
        Ok(())
    }

    async fn disconnect(&mut self) -> Result<()> {
        // matrix-sdk doesn't have an explicit disconnect; dropping the sync task suffices
        self.connected = false;
        Ok(())
    }

    fn on_message(&mut self, handler: MessageHandler) {
        let h = Arc::clone(&self.handler);
        tokio::task::block_in_place(|| {
            let rt = tokio::runtime::Handle::current();
            rt.block_on(async {
                let mut guard = h.lock().await;
                *guard = Some(handler);
            });
        });
    }

    async fn send(&self, msg: OutboundMessage) -> Result<Option<String>> {
        let client = self.client.as_ref().ok_or_else(|| {
            Error::Send("not connected".to_string())
        })?;

        let room_id: OwnedRoomId = RoomId::parse(&msg.chat_id)
            .map_err(|e| Error::Send(format!("invalid room_id: {}", e)))?;

        let room = client
            .get_room(&room_id)
            .ok_or_else(|| Error::Send(format!("room not found: {}", msg.chat_id)))?;

        let content = RoomMessageEventContent::text_plain(&msg.text);
        let response = room
            .send(content)
            .await
            .map_err(|e| Error::Send(format!("matrix send: {}", e)))?;

        Ok(Some(response.event_id.to_string()))
    }

    async fn get_status(&self) -> ChannelStatus {
        ChannelStatus {
            connected: self.connected,
            channel: "matrix".to_string(),
            account_id: self.user_id.clone(),
            error: None,
            last_activity: self.last_activity,
        }
    }
}
